### ISSUE TITLE
more rpki brokenness longer objects vs shorter RPKI.

### ISSUE THREAD

**Issue by troy2914** [irrdnet/irrd#465]
seen with 4.1.2
looking up 210.130.209.0/24
```
route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR
rpki-ov-state:  not_found # No ROAs found, or RPKI validation not enabled for source

route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        w-ippei@iij.ad.jp 20141010
source:         RADB
rpki-ov-state:  not_found # No ROAs found, or RPKI validation not enabled for source

route:          210.130.0.0/16
descr:          RPKI ROA for 210.130.0.0/16 / AS2497
remarks:        This AS2497 route object represents routing data retrieved
                from the RPKI. This route object is the result of an automated
                RPKI-to-IRR conversion process performed by IRRd.
max-length:     16
origin:         AS2497
source:         RPKI  # Trust Anchor: apnic

route:          210.130.209.0/24
descr:          Prolexic Customer Route
origin:         AS32787
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR
rpki-ov-state:  not_found # No ROAs found, or RPKI validation not enabled for source

route:          210.130.209.0/24
descr:          Prolexic Customer Route
origin:         AS32787
mnt-by:         MAINT-AS2497
changed:        noc@iij.ad.jp 20151124
source:         RADB
rpki-ov-state:  not_found # No ROAs found, or RPKI validation not enabled for source
```

give the RPKI object of /16 with a max-length of 16,  why are those /24's not invalid?

the cloudfare rpki validator says the /24s should be invalid (and I agree):
https://rpki.cloudflare.com/?view=validator&validateRoute=32787_210.130.209.0%2F24

**Comment by troy2914** [comment]
Hmm,  these seem to have cleaned up at some point, multiple hours later.
```
route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR
rpki-ov-state:  valid

route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        w-ippei@iij.ad.jp 20141010
source:         RADB
rpki-ov-state:  valid

route:          210.130.0.0/16
descr:          RPKI ROA for 210.130.0.0/16 / AS2497
remarks:        This AS2497 route object represents routing data retrieved
                from the RPKI. This route object is the result of an automated
                RPKI-to-IRR conversion process performed by IRRd.
max-length:     16
origin:         AS2497
source:         RPKI  # Trust Anchor: apnic
```

**Comment by troy2914** [comment]
note the upgrade to 4.1.2 was at 14:20,  and this was opened around 21:30

**Comment by mxsasha** [comment]
That is intriguing. I will look into it :)

**Comment by troy2914** [comment]
and i first noticed it at 16:30

**Comment by mxsasha** [comment]
At this point I'm thinking that this is was a hangover from #461. It was fixed without action, and I can't reproduce it here. And this is precisely the issue from #461, given that you said at that time that RADB had rpki_excluded set:

- 210.130.0.0/16 / AS2497 exists in two sources, you said one (RADB) had rpki_excluded set. The correct states would therefore be JPIRR valid, RADB not_found, but they are both not_found.
- 210.130.209.0/24 / AS32787 exists in the same two sources. Correct states would be JPIRR invalid, RADB not_found (because it's excluded), but they are both not_found.

If there was an issue regarding more specific objects, this would likely be a much wider issue. I have run checks against some of Job's unrelated validation code at the time of development and did not see this. Also, if there was a problem around more specifics, the states for the /16 would be JPIRR valid and RADB not_found, instead of 2x not_found. So considering that this meets the circumstances of #461 precisely and was transient, I do not think this is a second, different bug.

**Comment by troy2914** [comment]
actually when i upgraded to 4.1.2 I pulled in change where RADB was not rpki_excluded.

**Comment by troy2914** [comment]
as you noticed there are two bugs with the original query,  showing the 2 as not_found when they matched the RPKI they should have been valid.   And the 2 /24's making it thru.

**Comment by troy2914** [comment]
also I had queried for invalid's at 16:21 on the 10th,  so about 2 hours after the upgrade.
with
```
 psql -d irrd -t -A -q -c "select rpsl_pk,rpki_status,source from  rpsl_objects where rpki_status = 'invalid';"
```

That resulted in 48,786 objects.   doing that query now results in 203,704 objects.

**Comment by troy2914** [comment]
are there anything in the server.log when it re-authenticates objects?

**Comment by troy2914** [comment]
```
irrd@irrd01-beta:~$ grep RADB rpki.invalid.20210210.1621.txt |wc -l
3110
irrd@irrd01-beta:~$ grep RADB rpki.invalid.20210212.2356.txt |wc -l
157860
irrd@irrd01-beta:~$ grep JPIRR rpki.invalid.20210210.1621.txt |wc -l
32
irrd@irrd01-beta:~$ grep JPIRR rpki.invalid.20210212.2356.txt |wc -l
41
irrd@irrd01-beta:~$ grep NTTCOM rpki.invalid.20210210.1621.txt |wc -l
243
irrd@irrd01-beta:~$ grep NTTCOM rpki.invalid.20210212.2356.txt |wc -l
0
```

NTTCOM does have rpki_excluded in the irrd.yaml.

where those filenames are based on 
```
rpki.invalid.`date +%Y%m%d.%H%M`.txt
```

**Comment by troy2914** [comment]
complete list of changes:
```
irrd@irrd01-beta:~$ cut -d\|  -f3 rpki.invalid.20210210.1621.txt | sort | uniq -c
     77 AFRINIC
    109 ALTDB
   9020 APNIC
    531 ARIN
   1985 ARIN-NONAUTH
   5284 ARIN-WHOIS
     24 BBOI
     77 BELL
     32 JPIRR
     12 LACNIC
   5440 LEVEL3
    243 NTTCOM
   3110 RADB
     18 REGISTROBR
      2 RGNET
  22735 RIPE
      6 RIPE-NONAUTH
     81 TC
irrd@irrd01-beta:~$ cut -d\|  -f3 rpki.invalid.20210212.2356.txt | sort | uniq -c
     80 AFRINIC
    109 ALTDB
   9110 APNIC
    537 ARIN
   1990 ARIN-NONAUTH
   5306 ARIN-WHOIS
     26 BBOI
     77 BELL
     41 JPIRR
      8 LACNIC
   5487 LEVEL3
 157860 RADB
     18 REGISTROBR
      2 RGNET
  22962 RIPE
     10 RIPE-NONAUTH
     81 TC
```

**Comment by mxsasha** [comment]
> as you noticed there are two bugs with the original query, showing the 2 as not_found when they matched the RPKI they should have been valid. And the 2 /24's making it thru.

Yes, but that is exactly the kind of bug that #461 would trigger in one of it's states: incorrectly marking objects as not_found. Which is why it seems odd that another overlapping bug would occur at the same time. But not impossible.

> are there anything in the server.log when it re-authenticates objects?

The log messages are:
1. Running full ROA import from: {roa_source}, SLURM {slurm_source}
2. ROA import from {roa_source}, SLURM {slurm_source}, imported {len(roa_importer.roa_objs)} ROAs, running validator
3. RPKI status updated for all routes, ...

After 2, the ROA table is updated, which means it will be used for NRTM, auth changes, etc. once that has been committed. The third line is when the validator has been run against all objects that are currently in the database, and the result has been committed. (There are indeed two commit steps: one after line 2, the second before line 3.)

**Comment by troy2914** [comment]
interesting,  I was just looking at the server.log's that I saved:
2021-02-10 14:22:26,789 irrd[90730]: [irrd.mirroring.mirror_runners_import#INFO] Running full ROA import from: https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json
(last log in this file is 2021-02-11 00:00:04,094)
2021-02-11 09:05:49,213 irrd[125574]: [irrd.mirroring.mirror_runners_import#INFO] Running full ROA import from: https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json
(first one in that is 2021-02-11 00:00:21,428)

before that 2/10 date it was hourly, and on 2/11 after that date it was hourly.

**Comment by troy2914** [comment]
i just went back to 4.0.8 to redo the upgrade.
pre-starting up (but post db upgrade):
pre-start condition (that had a RPKI manual source):
```
210.130.0.0/16AS2497|not_found|RPKI
210.130.0.0/16AS2497|not_found|JPIRR
210.130.0.0/16AS2497|not_found|RADB
210.130.209.0/24AS32787|not_found|JPIRR
210.130.209.0/24AS32787|not_found|RADB
```

start up at Sat 2021-02-20 07:38:18 UTC
1st 
2021-02-20 07:38:19,332 irrd[55416]: [irrd.mirroring.mirror_runners_import#INFO] Running full ROA import from: https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json
1 second after startup, hmm.
it was still
```
210.130.0.0/16AS2497|not_found|RPKI
210.130.0.0/16AS2497|not_found|JPIRR
210.130.0.0/16AS2497|not_found|RADB
210.130.209.0/24AS32787|not_found|JPIRR
210.130.209.0/24AS32787|not_found|RADB
```

a few minutes go by, and the RPKI entry goes away:
```
210.130.0.0/16AS2497|not_found|JPIRR
210.130.0.0/16AS2497|not_found|RADB
210.130.209.0/24AS32787|not_found|JPIRR
210.130.209.0/24AS32787|not_found|RADB
```
2021-02-20 07:44:23,598 irrd[55416]: [irrd.mirroring.mirror_runners_import#INFO] ROA import from https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json, imported 219817 ROAs, running validator

still showing 4 not_found entries at 08:00
waiting for the next full ROA import.   a `whois -h localhost 210.130.209.0/24`
is returning "Timeout."  (at 7:55, and again at 8:01, over 20 minutes from the start of the irrd process)

is there a way for force a full ROA import?   And what is the message when the import validator ends?

**Comment by troy2914** [comment]
sigh,  
```
2021-02-20 08:40:31,095 irrd[55416]: [irrd.mirroring.mirror_runners_import#ERROR] An exception occurred while attempting a ROA import: (psycopg2.errors.DeadlockDetected) deadlock detected
```

**Comment by troy2914** [comment]
seems the following ROA import's where succesful:
```
2021-02-20 08:40:39,977 irrd[59322]: [irrd.mirroring.mirror_runners_import#INFO] Running full ROA import from: https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json
2021-02-20 08:46:57,442 irrd[59322]: [irrd.mirroring.mirror_runners_import#INFO] ROA import from https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json, imported 219831 ROAs, running validator
2021-02-20 11:33:25,904 irrd[75456]: [irrd.mirroring.mirror_runners_import#INFO] Running full ROA import from: https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json
2021-02-20 11:38:57,474 irrd[75456]: [irrd.mirroring.mirror_runners_import#INFO] ROA import from https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json, imported 219841 ROAs, running validator
2021-02-20 12:33:31,498 irrd[98043]: [irrd.mirroring.mirror_runners_import#INFO] Running full ROA import from: https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json
```

irrd 4.1.2 started at Sat 2021-02-20 07:38:18 UTC
The `whois -h localhost 210.130.209.0/24` started responding at ~09:45 with RPKI valid, others being not_found
then at about ~11:33 only 3 valid /16 started being returned.

I am curious why that whois was timing out between 07:38 and 09:45,  and if there are any log messages when the validation ends.

**Comment by mxsasha** [comment]
I am having a bit of trouble reading what the current open concerns are. Both on your test instance and mine, I currently see the following objects for 210.130.209.0/24, **if I disable the RPKI filter for the query** (`!fno-rpki-filter`):

```
route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        w-ippei@iij.ad.jp 20141010
source:         RADB
rpki-ov-state:  valid

route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR
rpki-ov-state:  valid

route:          210.130.0.0/16
descr:          RPKI ROA for 210.130.0.0/16 / AS2497
remarks:        This AS2497 route object represents routing data retrieved
                from the RPKI. This route object is the result of an automated
                RPKI-to-IRR conversion process performed by IRRd.
max-length:     16
origin:         AS2497
source:         RPKI  # Trust Anchor: apnic

route:          210.130.209.0/24
descr:          Prolexic Customer Route
origin:         AS32787
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR
rpki-ov-state:  invalid

route:          210.130.209.0/24
descr:          Prolexic Customer Route
origin:         AS32787
mnt-by:         MAINT-AS2497
changed:        noc@iij.ad.jp 20151124
source:         RADB
rpki-ov-state:  invalid
```

So the current state is entirely correct - or at least, consistent with the translated ROA in the RPKI source. I'm not entirely following what the remaining concerns are.

> is there a way for force a full ROA import?

That is already done every RPKI run. To be precise, the RPKI import process works as follows:

1. The scheduler starts the ROAImportRunner whenever rpki.roa_source is set (has a default), and rpki.roa_import_timer (default 1 hour) has expired since the last run, and there is no ROAImportRunner already running. The timer is also considered expired when IRRd starts. If an importer is already running when the timer expires, the next run is started after the current one finishes. The scheduler checks all this every 15 seconds.
2. The importer starts a new transaction. Database changes are not visible to other connections until commit.
3. **The importer logs "Running full ROA import from: {roa_source}, SLURM {slurm_source}"**
4. The importer deletes everything from roa_object and deletes all rpsl_object objects with source RPKI
5. The ROA JSON file and SLURM file are retrieved/downloaded/gunzipped.
6. The import runner starts ROADataImporter with the ROA and SLURM data.
7. The ROADataImporter creates entries for the roa_object and rpsl_object tables. These are buffered to some degree in the database handler. The importer also keeps all ROA objects in memory, for use by the validator later.
8. **The import runner logs "ROA import from {roa_source}, SLURM {slurm_source}, imported {len(roa_importer.roa_objs)} ROAs, running validator"**
9. ROAImportRunner commits the database transaction. This also flushes all buffers. New RPKI RPSL and ROA objects are now visible to other connections. A new transaction is started.
10. The import runner calls BulkRouteROAValidator.
11. BulkRouteROAValidator loads the ROA objects in a trie for fast searching.
12. BulkRouteROAValidator reads all route(6) objects from rpsl_object, and validates each of them against the ROAs. The validation result from sources with rpki_excluded set, is always not_found.
13. For all validated objects, BulkRouteROAValidator checks whether their validation state recorded in rpsl_objects, matches the validation result based on the current ROAs.
14. All objects that need a validation state update are returned.
15. The import runner passes the objects that need an updated validation state to the database handler.
16. The database handler updates these records in bulk, based on their PK (not the RPSL PK - that was the bug in #461, but the unique UUID pk column in each row).
17. The database handler runs through each state change, and any object that changed to or from invalid, generates a ADD/DEL journal entry if enabled for the source.
18. The transaction is committed.
19. Owners of newly invalid objects from authoritative sources are notified, if enabled.
20. **The import runner logs "RPKI status updated for all routes, {len(objs_now_valid)} newly valid, ..."**
21. The importer is finished.

When an NRTM update or authoritative change occurs, the RPKI status on insertion/update is set by searching the roa_object table directly.

**Comment by troy2914** [comment]
4. explains why I saw the RPKI source go away.   Though that would imply that action was _not_ done in the transaction that was opened at 2.  That seems like a bug?
or the failed update at 
```
2021-02-20 08:40:31,095 irrd[55416]: [irrd.mirroring.mirror_runners_import#ERROR] An exception occurred while attempting a ROA import: (psycopg2.errors.DeadlockDetected) deadlock detected
```
resulted in the transaction with the removal of the RPKI to be commited,  which again sounds like a bug?

and, i see the RPKI status end message at:
```
2021-02-20 11:31:26,704 irrd[59322]: [irrd.mirroring.mirror_runners_import#INFO] RPKI status updated for all routes,
888392 newly valid, 203634 newly invalid, 0 newly not_found routes, 0 emails sent to contacts of newly invalid author
itative objects
```
Which explains why at the 11:33 query (run every 3 minutes),  it showed the correct results.  And also why no roa import where run between 08:44 and 11:32.

And is a 2 hour period expected between the new upgraded irrd starting,  and it responding to whois queries?

**Comment by mxsasha** [comment]
I see it changed the status for 1 million objects, so it makes sense that it can take a while. That is a very exceptional case as well.

As far as I understand there are now two open concerns:
* Transaction management in the RPKI importer may not work correctly, as the rpsl_objects table did not contain any objects with source RPKI at a point in time.
* Whois queries may be blocked during some parts of the RPKI importer.

I'm going to try to reproduce that and dig into it more, but I first want to make sure I understand the scenario you had. From what I understand you:
1. Ran 4.1.2 for a while
2. Downgraded to 4.0.8, started it and ran it for a while
3. Upgraded to 4.1.2 again and started it
4. And, at no point in this process you ran any database migrations with irrd_database_upgrade/downgrade, right?

If I understand you right, the two concerns appeared right after step 3, right?

(On a sidenote, I should say that downgrading/upgrading between 4.0.x and 4.1.x without running the migrations is not a supported setup, and can cause issues - it can cause route objects to miss their prefix_length in the rpsl_objects table, which is also used by the importer. However, I don't see that as likely to be related to the two concerns I summarised above.)

**Comment by troy2914** [comment]
for step 2 the downgrade to 4.0.8 that also included
a. reverting the service file
b. reverting the yaml config file
c. loading a pg_dumpall from another host
d. pip install irrd==4.0.8
e. systemctl start irrd

then I ran puppet, which installed irrd==4.1.2,  started a irrd_database_upgrade that failed
due it to taking so long.
re-ran irrd_database_upgrade (which took about 45 minutes)
during that process irrd was started by puppet,
i stopped it,  then disabled puppet

yes , the considers started right after step 3. starting irrd 4.1.2

**Comment by troy2914** [comment]
interesting i just repeated stopping irrd 4.1.2
loading a pg_dumpall from another host (4.0.8 db)
and running the irrd_database_upgrade
starting irrd 4.1.2  time Sun 2021-02-21 20:15:28 UTC
and the first ROA message at:
```
2021-02-21 20:15:29,822 irrd[51839]: [irrd.mirroring.mirror_runners_import#INFO] Running full ROA import from: https://rpki.gin.ntt.net/api/export.json, SLURM file:///etc/rpki-slurm.json
```
but now the whois is working from the start, so that's good.   err no.
```
-rw-r--r-- 1 irrd irrd 1535 Feb 21 20:16 stats.20210221-201601
-rw-r--r-- 1 irrd irrd    0 Feb 21 20:18 stats.20210221-201801
-rw-r--r-- 1 irrd irrd    0 Feb 21 20:20 stats.20210221-202001
-rw-r--r-- 1 irrd irrd    0 Feb 21 20:22 stats.20210221-202201
-rw-r--r-- 1 irrd irrd    0 Feb 21 20:24 stats.20210221-202401
```
the 1st query after start worked, but now they are timing out.
seems to be back to working at Feb 21 21:02

but between
Sun 21 Feb 2021 08:20:31 PM UTC
and
Sun 21 Feb 2021 08:20:34 PM UTC
my RPKI source route was removed
and the validator has not completed yet.
where that was seen via:  
```
psql -d irrd -t -A -q -c "select rpsl_pk,rpki_status,source from  rpsl_objects where rpsl_pk in ('210.130.0.0/16AS2497','210.130.209.0/24AS32787');
```

and the validation completed at:
```
2021-02-21 22:49:45,778 irrd[51839]: [irrd.mirroring.mirror_runners_import#INFO] RPKI status updated for all routes, 888365 newly valid, 203807 newly invalid, 0 newly not_found routes, 0 emails sent to contacts of newly invalid authoritative objects
```
and the whois is now correct with:
```
route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        w-ippei@iij.ad.jp 20141010
source:         RADB
rpki-ov-state:  valid

route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR
rpki-ov-state:  valid

route:          210.130.0.0/16
descr:          RPKI ROA for 210.130.0.0/16 / AS2497
remarks:        This AS2497 route object represents routing data retrieved
                from the RPKI. This route object is the result of an automated
                RPKI-to-IRR conversion process performed by IRRd.
max-length:     16
origin:         AS2497
source:         RPKI  # Trust Anchor: apnic
```

oh, interesting,  rpki source is still missing:
```
210.130.0.0/16AS2497|valid|JPIRR
210.130.0.0/16AS2497|valid|RADB
210.130.209.0/24AS32787|invalid|JPIRR
210.130.209.0/24AS32787|invalid|RADB
```

So I guess that it is removed early does not matter.   And that was just cleanup from the 4.0.8 db, vs the 4.1.2 yaml (where there is no explicit rpki source).

Therefore the only outstanding issue, is the timeout on the whois queries.

**Comment by mxsasha** [comment]
I ran some very thorough testing on this, and kept a close eye on the database status, locks, queries and whois output. This is a very long comment, also for my own notes, conclusions at the end.

On a test machine, I made a fresh install of 4.0.8 with a fresh database. I had it import all the usual sources, and let it run for a while. I see the following objects in whois:
```
# whois -h localhost 210.130.209.0/24
route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR

route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        w-ippei@iij.ad.jp 20141010
source:         RADB

route:          210.130.209.0/24
descr:          Prolexic Customer Route
origin:         AS32787
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR

route:          210.130.209.0/24
descr:          Prolexic Customer Route
origin:         AS32787
mnt-by:         MAINT-AS2497
changed:        noc@iij.ad.jp 20151124
source:         RADB
```

Which matches what I see in the SQL database:
```
irrd=# SELECT rpsl_pk, source FROM rpsl_objects WHERE ip_first <= '210.130.209.0' AND ip_last >= '210.130.209.255';
         rpsl_pk         | source
-------------------------+--------
 210.130.209.0/24AS32787 | RADB
 210.130.209.0/24AS32787 | JPIRR
 210.130.0.0/16AS2497    | RADB
 210.130.0.0/16AS2497    | JPIRR
```
I did not have an "old style" RPKI source configured, so it is correct that there is no RPKI object.

Then, I stopped IRRd, upgraded to 4.1.2, and then ran irrd_database_upgrade (22 minutes). Before starting, the database says:
```
# SELECT rpsl_pk, source, rpki_status FROM rpsl_objects WHERE ip_first <= '210.130.209.0' AND ip_last >= '210.130.209.255';
         rpsl_pk         | source | rpki_status
-------------------------+--------+-------------
 210.130.209.0/24AS32787 | JPIRR  | not_found
 210.130.209.0/24AS32787 | RADB   | not_found
 210.130.0.0/16AS2497    | JPIRR  | not_found
 210.130.0.0/16AS2497    | RADB   | not_found
 ```
This is expected - not_found is the default internal initial state.

This query allows inspecting current locks:
```
SELECT a.datname, l.relation::regclass, l.transactionid, l.mode, l.GRANTED, a.usename, a.query, a.query_start, age(now(), a.query_start) AS "age", a.pid
FROM pg_stat_activity a
JOIN pg_locks l ON l.pid = a.pid
ORDER BY a.query_start;
```
and showed that PostgreSQL had started an autovacuum with VACUUM ANALYZE. That is normal, because the table has changed substantially in the migration. While this is running, SELECT queries are not blocked, both confirmed by trying out and by looking at the lock modes. The VACUUM acquires ROW EXCLUSIVE, which does not conflict with ACCESS SHARE acquired by SELECT.

I started IRRd after a while, which immediately started the RPKI importer. The RPKI importer is not locked by the VACUUM. Whois queries continue to work in the first and second stage of the RPKI importer.

After the first stage finishes and is committed, the new RPKI object appears, but validation state is not updated yet, as expected (I had not enabled RPKI in the default sources, hence two queries):
```
$ whois -h localhost 210.130.209.0/24
route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR
rpki-ov-state:  not_found # No ROAs found, or RPKI validation not enabled for source

route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        w-ippei@iij.ad.jp 20141010
source:         RADB
rpki-ov-state:  not_found # No ROAs found, or RPKI validation not enabled for source

route:          210.130.209.0/24
descr:          Prolexic Customer Route
origin:         AS32787
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR
rpki-ov-state:  not_found # No ROAs found, or RPKI validation not enabled for source

route:          210.130.209.0/24
descr:          Prolexic Customer Route
origin:         AS32787
mnt-by:         MAINT-AS2497
changed:        noc@iij.ad.jp 20151124
source:         RADB
rpki-ov-state:  not_found # No ROAs found, or RPKI validation not enabled for source

$ whois -h localhost -s RPKI 210.130.209.0/24
Warning: RIPE flags used with a traditional server.
route:          210.130.0.0/16
descr:          RPKI ROA for 210.130.0.0/16 / AS2497
remarks:        This AS2497 route object represents routing data retrieved
                from the RPKI. This route object is the result of an automated
                RPKI-to-IRR conversion process performed by IRRd.
max-length:     16
origin:         AS2497
source:         RPKI  # Trust Anchor: apnic
```
**Important note**: the first whois query here was incredibly slow, about 43 seconds. More on that later. 

I also see it in the SQL database as expected:
```
# SELECT rpsl_pk, source, rpki_status FROM rpsl_objects WHERE ip_first <= '210.130.209.0' AND ip_last >= '210.130.209.255';
          rpsl_pk          | source | rpki_status
---------------------------+--------+-------------
 210.130.209.0/24AS32787   | JPIRR  | not_found
 210.130.209.0/24AS32787   | RADB   | not_found
 210.130.0.0/16AS2497/ML16 | RPKI   | valid
 210.130.0.0/16AS2497      | JPIRR  | not_found
 210.130.0.0/16AS2497      | RADB   | not_found
 ```
(Objects from the RPKI source itself are always internally marked as valid.)

At some point, I started seeing the UPDATE queries from the RPKI importer to update the RPKI status. These are big, because all valid and invalid objects are set to not_found after the migration, so there are many objects to update. Also during this stage, whois queries do not block, and nothing in the locks suggests that RPKI UPDATE, VACUUM and SELECT for whois are interfering.

At some point during this, after about 40 minutes, the VACUUM finishes. This makes sense, because rpsl_objects has quite a few indexes, needed to support all the different query types.

After the vacuum finishes, the `whois -h localhost 210.130.209.0/24` query is fast again: less than one second. Comparing the EXPLAIN ANALYSE output explains why:
```
# Before VACUUM finishes:
irrd=# explain analyse SELECT rpsl_objects.pk, rpsl_objects.object_class, rpsl_objects.rpsl_pk, rpsl_objects.parsed_data, rpsl_objects.object_text, rpsl_objects.source, rpsl_objects.rpki_status FROM rpsl_objects WHERE rpsl_objects.source IN ('NTTCOM', 'RADB', 'RIPE', 'RIPE-NONAUTH', 'ALTDB', 'BELL', 'LEVEL3', 'RGNET', 'APNIC', 'JPIRR', 'ARIN', 'ARIN-NONAUTH', 'BBOI', 'TC', 'AFRINIC') AND rpsl_objects.rpki_status IN ('not_found', 'valid') AND rpsl_objects.scopefilter_status IN ('in_scope') AND rpsl_objects.ip_first <= '210.130.209.0' AND rpsl_objects.ip_last >= '210.130.209.255' AND rpsl_objects.ip_version = 4 ORDER BY rpsl_objects.ip_first ASC, rpsl_objects.asn_first ASC, rpsl_objects.rpsl_pk ASC;
                                                                                                                            QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=810.55..810.58 rows=14 width=765) (actual time=45516.523..45516.525 rows=4 loops=1)
   Sort Key: ip_first, asn_first, rpsl_pk
   Sort Method: quicksort  Memory: 27kB
   ->  Bitmap Heap Scan on rpsl_objects  (cost=607.24..810.28 rows=14 width=765) (actual time=4252.502..45515.636 rows=4 loops=1)
         Recheck Cond: ((scopefilter_status = 'in_scope'::scopefilterstatus) AND (rpki_status = ANY ('{not_found,valid}'::rpkistatus[])))
         Filter: ((ip_first <= '210.130.209.0'::inet) AND (ip_last >= '210.130.209.255'::inet) AND (ip_version = 4) AND ((source)::text = ANY ('{NTTCOM,RADB,RIPE,RIPE-NONAUTH,ALTDB,BELL,LEVEL3,RGNET,APNIC,JPIRR,ARIN,ARIN-NONAUTH,BBOI,TC,AFRINIC}'::text[])))
         Rows Removed by Filter: 3938316
         Heap Blocks: exact=427077
         ->  BitmapAnd  (cost=607.24..607.24 rows=195 width=0) (actual time=2228.702..2228.702 rows=0 loops=1)
               ->  Bitmap Index Scan on ix_rpsl_objects_scopefilter_status  (cost=0.00..202.33 rows=19453 width=0) (actual time=1074.646..1074.646 rows=3950746 loops=1)
                     Index Cond: (scopefilter_status = 'in_scope'::scopefilterstatus)
               ->  Bitmap Index Scan on ix_rpsl_objects_rpki_status  (cost=0.00..404.66 rows=38905 width=0) (actual time=954.323..954.323 rows=3951790 loops=1)
                     Index Cond: (rpki_status = ANY ('{not_found,valid}'::rpkistatus[]))
 Planning time: 2.584 ms
 Execution time: 45517.288 ms
 
 # After VACUUM finishes (RPKI importer still running in phase 2 with a huge UPDATE):
 irrd=# explain analyse SELECT rpsl_objects.pk, rpsl_objects.object_class, rpsl_objects.rpsl_pk, rpsl_objects.parsed_data, rpsl_objects.object_text, rpsl_objects.source, rpsl_objects.rpki_status FROM rpsl_objects WHERE rpsl_objects.source IN ('NTTCOM', 'RADB', 'RIPE', 'RIPE-NONAUTH', 'ALTDB', 'BELL', 'LEVEL3', 'RGNET', 'APNIC', 'JPIRR', 'ARIN', 'ARIN-NONAUTH', 'BBOI', 'TC', 'AFRINIC') AND rpsl_objects.rpki_status IN ('not_found', 'valid') AND rpsl_objects.scopefilter_status IN ('in_scope') AND rpsl_objects.ip_first <= '210.130.209.0' AND rpsl_objects.ip_last >= '210.130.209.255' AND rpsl_objects.ip_version = 4 ORDER BY rpsl_objects.ip_first ASC, rpsl_objects.asn_first ASC, rpsl_objects.rpsl_pk ASC;
                                                                                                                                              QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=376357.72..377322.17 rows=385779 width=750) (actual time=135.392..135.395 rows=4 loops=1)
   Sort Key: ip_first, asn_first, rpsl_pk
   Sort Method: quicksort  Memory: 27kB
   ->  Index Scan using ix_rpsl_objects_ip_last_ip_first on rpsl_objects  (cost=0.43..340562.41 rows=385779 width=750) (actual time=0.312..135.366 rows=4 loops=1)
         Index Cond: ((ip_last >= '210.130.209.255'::inet) AND (ip_first <= '210.130.209.0'::inet))
         Filter: ((rpki_status = ANY ('{not_found,valid}'::rpkistatus[])) AND (scopefilter_status = 'in_scope'::scopefilterstatus) AND (ip_version = 4) AND ((source)::text = ANY ('{NTTCOM,RADB,RIPE,RIPE-NONAUTH,ALTDB,BELL,LEVEL3,RGNET,APNIC,JPIRR,ARIN,ARIN-NONAUTH,BBOI,TC,AFRINIC}'::text[])))
         Rows Removed by Filter: 1
 Planning time: 0.393 ms
 Execution time: 135.437 ms
 ```
This is an expected behaviour: analysing the table, done as part of the VACUUM ANALYZE, helps the query planner to figure out which indexes to use. After the significant changes from the migration, the query planner at first planned to scan the table, until analysing finished, and it used the index. To be clear, the VACUUM ANALYZE is a process automatically started when needed by PostgreSQL. Also, the impact on query planning is usually not this bad - that happened because of the one time migration..

**Important note**: with ~200.000 route objects being flipped from not_found to invalid, on the first import, this also generates 200.000 DELETE entries in the journal. This takes some time, and during this time processing NRTM updates and authoritative changes may be blocked. This is by design, due to difficult complications around NRTM serial handling. After the first import, the volume of journal changes is small, and this is not a concern. But it might be noticeable on the first RPKI import. Blocked changes will be processed afterwards.

About 35 minutes after starting IRRd, the first RPKI importer completes, and status is correct:
```
2021-02-23 14:11:04,765 irrd[18232]: [irrd.mirroring.mirror_runners_import#INFO] RPKI status updated for all routes, 927866 newly valid, 239072 newly invalid, 0 newly not_found routes, 0 emails sent to contacts of newly invalid authoritative objects

# SELECT rpsl_pk, source, rpki_status FROM rpsl_objects WHERE ip_first <= '210.130.209.0' AND ip_last >= '210.130.209.255';
          rpsl_pk          | source | rpki_status
---------------------------+--------+-------------
 210.130.209.0/24AS32787   | RADB   | invalid
 210.130.209.0/24AS32787   | JPIRR  | invalid
 210.130.0.0/16AS2497      | RADB   | valid
 210.130.0.0/16AS2497      | JPIRR  | valid
 210.130.0.0/16AS2497/ML16 | RPKI   | valid
 
 $ whois -h localhost 210.130.209.0/24
route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        w-ippei@iij.ad.jp 20141010
source:         RADB
rpki-ov-state:  valid

route:          210.130.0.0/16
descr:          IIJ IPv4 BLOCK ( AS2497 )
origin:         AS2497
mnt-by:         MAINT-AS2497
changed:        shuhei-o@iij.ad.jp 20200716
source:         JPIRR
rpki-ov-state:  valid
```
(RPKI source was not enabled by default, hence two, not three, objects returned from whois.)

The big changes made by the initial RPKI import triggered another automatic vacuum, which is expected. During this vacuum, whois continues to perform normally, using the right indexes.

After the vacuum finished, I restarted IRRd to force another RPKI import. During this, whois queries were normal. Phase 1 took 4 minutes, phase 2 as well. Phase 1 did trigger another vacuum, but it was a very short one.

During phase 1, where all RPKI RPSL objects are deleted briefly in a transaction, I also checked the rpsl_objects table and whois output, and consistently did find all expected objects, also after the (in transaction) DELETE happened. So transaction isolation was functioning as intended.

### Conclusions

When thoroughly monitoring all steps of the 4.0.8->4.1.2 upgrade path, my testing shows no indication of data being unexpectedly incomplete or incorrect, transaction isolation not functioning, or whois queries being entirely blocked, at any point. Shortly after migration, queries may be very slow for a short while. Specifically:

* During the upgrade path, in the final part of phase 2 (validating existing objects) of the RPKI importer, NRTM updates and authoritative changes are blocked, but resumed later. This is a known limitation, and not significant in later runs which update a smaller number of objects.
* Except for this known limitation, there seem to be no processes interfering with each other due to locking, determined by both manual checking and reviewing the active locks.
* After the 4.1 database upgrade, prior to the first automatic post-migration VACUUM ANALYSE finishing, the PostgreSQL query planner has difficulty determining the right indexes, and some whois queries may be incredibly slow, as they execute full table scans. A simple query for a /24, matching 4 objects, took 45 seconds in my testing setup, which may make it seem like IRRd is unresponsive and can cause timeouts in some clients. This is automatically resolved as PostgreSQL's autovacuum recognises the need for a VACUUM ANALYSE and starts it, but it can take in the order of half an hour for this to finish. Regular running of the RPKI importer does not cause these kind of query planner issues - it's most likely the database migration that is the original cause.

**Comment by troy2914** [comment]
Ok, well I'm not sure what the default timeout on my debian 10.3 whois client is.   I guess everything is resolved.
Closing this issue.
